Translate to Brazilian Portuguese the containers section of the basic training (#175)
nextflow-io/training: docs/basic_training/containers.pt.md

---
description: Material de treinamento básico do Nextflow
---

# Gerencie dependencias e containers

Fluxos de trabalhos computacionais são raramente compostos de um só script ou ferramenta.  Muitas vezes, eles dependem de duzias de componentes de softwares ou bibliotecas.

Instalar e manter tais dependências é uma tarefa desafiadora e uma fonte comum de irreprodutibilidade em aplicações científicas.

> Para superar esses problemas, nós utilizamos contêineres que habilitam essas dependências de softwares, isto é ferramentas e bibliotecas necessárias para uma análise de dados, para estar encapsuladas em uma ou mais imagens de contêiner Linux independentes, prontas para serem executadas e imutáveis. Elas podem facilmente ser implementadas em qualquer plataforma que suporta o motor de conteinerização.

Reviewer: ```suggestion
Para superar esses problemas, nós utilizamos contêineres que habilitam essas dependências de softwares, isto é ferramentas e bibliotecas necessárias para uma análise de dados, para estar encapsuladas em uma ou mais imagens de contêiner Linux independentes, prontas para serem executadas e imutáveis. Elas podem facilmente ser implementadas em qualquer plataforma que suporta o motor de contêinerização.
```

```bash
docker run hello-world
```

### Baixe um contêiner

O comando pull possibilita que você baixe uma imagem Docker sem que a execute. Por exemplo:

```bash
docker pull debian:stretch-slim
```

O comando acima baixa uma imagem Debian Linux. Você pode checar se ela existe usando:

```bash
docker images
```

### Execute um contêiner em mode interativo

Iniciar uma shell BASH em um contêiner permite que você opere em modo interativo no sistema operacional contêinerizado. Por exemplo:

```
docker run -it debian:stretch-slim bash
```

Uma vez iniciado, você vai notar que está como root (!). Use os comandos usuais para navegar pelo sistema de arquivos. Isso é útil para checar se os programas necessários estão presentes no contêiner.

Para sair do contêiner, pare a sessão BASH com o comando `exit`.

### Seu primeiro Dockerfile

Imagens Docker são criadas utilizando um arquivo chamado `Dockerfile`, que é um simples arquivo de texto contendo uma lista de comandos para montar e configurar uma imagem com os pacotes de programas necessários.

Aqui, você criará uma imagem Docker contendo o cowsay e a ferramenta Salmon

!!! warning

>     O processo de montagem do Docker automaticamente copia todos arquivos que estão no diretório atual para o Docker daemon para que possa criar a imagem. Isso pode custar muito tempo quando existem vários ou grandes arquivos. Por essa razão, é importante que _sempre_ trabalhe em um diretório contendo apenas os arquivos que você realmente precisa incluir em sua imagem Docker. Alternativamente, você pode usar o arquivo `.dockerignore` para selecionar os aquivos que serão excluidos da montagem.

Reviewer: ```suggestion
    O processo de montagem do Docker automaticamente copia todos os arquivos que estão no diretório atual para o Docker daemon para que ele possa criar a imagem. Isso pode custar muito tempo quando existem vários ou grandes arquivos. Por essa razão, é importante que _sempre_ se trabalhe em um diretório contendo apenas os arquivos que você realmente precisa incluir em sua imagem Docker. Alternativamente, você pode usar o arquivo `.dockerignore` para selecionar os aquivos que serão excluídos da montagem.
```

```

### Monte a imagem

Monte a imagem Dockerfile utilizando o seguinte comando:

```bash
docker build -t my-image .
```

Onde "my-image" é o nome que o usuário especificou para Dockerfile, presente no diretório atual.

!!! tip

    Não esqueça do ponto no comando acima.

Quando completo, verifique se a imagem foi criada listando todas imagens disponíveis:

```bash
docker images
```

Você pode testar seu novo container executando esse comando:

```bash
docker run my-image cowsay Hello Docker!
```

### Adicione um pacote de programa a imagem

Adicione o pacote Salmon para a imagem Docker adicionando o seguinte trecho para o `Dockerfile`:

```dockerfile
RUN curl -sSL https://github.com/COMBINE-lab/salmon/releases/download/v1.5.2/salmon-1.5.2_linux_x86_64.tar.gz | tar xz \
&& mv /salmon-*/bin/* /usr/bin/ \
&& mv /salmon-*/lib/* /usr/lib/
```

Salve o arquivo e monte a imagem novamente utilizando o mesmo comando anterior:

```bash
docker build -t my-image .
```

Você perceberá que isso cria uma nova imagem Docker **mas** com um ID de imagem diferente.

### Execute Salmon no container

Cheque se Salmon está executando corretamente no container como mostrado abaixo:

```bash

> docker run my-image salmon --version

Reviewer: ```suggestion
docker run minha-imagem salmon --version
```

```

> Você pode até iniciar o container no modo interativo utilizando o seguinte comando:

Reviewer: ```suggestion
Você pode até iniciar o contêiner no modo interativo utilizando o seguinte comando:
```

```bash

> docker run -it my-image bash

Reviewer: ```suggestion
docker run -it minha-imagem bash
```

```

> Use o comando `exit` para finalizar a seção interativa.

Reviewer: ```suggestion
Use o comando `exit` para finalizar a sessão interativa.
```

### Montagem do sitema de arquivos

> Crie um arquivo índice de genoma utilizando Salmon no container.

Reviewer: ```suggestion
Crie um arquivo índice de genoma utilizando o Salmon no contêiner.
```

> ### Montagem do sitema de arquivos

Reviewer: ```suggestion
### Montagem do sistema de arquivos
```

> Tente executar Salmon no container com o seguinte comando:

Reviewer: ```suggestion
Tente executar o Salmon no container com o seguinte comando:
```

```bash

> docker run my-image \

Reviewer: ```suggestion
docker run minha-imagem \
```

    salmon index -t $PWD/data/ggal/transcriptome.fa -i transcript-index
```

> O comando acima falha porque Salmon não pode acessar o arquivo de entrada.

Reviewer: ```suggestion
O comando acima falha porque o Salmon não pode acessar o arquivo de entrada.
```

> Isso acontece porque o container executa em um sistema de arquivos totalmente diferente e não pode acessar o arquivo no sistema de arquivo do host por default.

Reviewer: ```suggestion
Isso acontece porque o contêiner executa em um sistema de arquivos totalmente diferente e não pode acessar o arquivo no sistema de arquivo do host por padrão.
```

> Você precisará usar a opção de linha de comando `--volume` para montar o(s) arquivo(s) de entrada por exemplo.

Reviewer: ```suggestion
Você precisará usar a opção de linha de comando `--volume` para montar o(s) arquivo(s) de entrada, por exemplo
```

```bash

> docker run --volume $PWD/data/ggal/transcriptome.fa:/transcriptome.fa my-image \

Reviewer: ```suggestion
docker run --volume $PWD/data/ggal/transcriptome.fa:/transcriptome.fa minha-imagem \
```

    salmon index -t /transcriptome.fa -i transcript-index
```

!!! warning

>     O `transcript-index` diretório gerado ainda está inacessível no sistema de arquivo do host.

Reviewer: ```suggestion
    O diretório `transcript-index` gerado ainda está inacessível no sistema de arquivo do sistema operacional hospedeiro.
```

> Um jeito mais fácil é montar o diretório original em um indêntico no container, isso permite que você utilize o mesmo caminho durante a execução dentro do container por exemplo.

Reviewer: ```suggestion
Um jeito mais fácil é montar o diretório original em um indêntico no contêiner. Isso permite que você utilize o mesmo caminho durante a execução dentro do contêiner, por exemplo
```

```bash

> docker run --volume $PWD:$PWD --workdir $PWD my-image \

Reviewer: ```suggestion
docker run --volume $PWD:$PWD --workdir $PWD minha-imagem \
```

> Um jeito mais fácil é montar o diretório original em um indêntico no contêiner. Isso permite que você utilize o mesmo caminho durante a execução dentro do contêiner, por exemplo

Reviewer: ```suggestion
Um jeito mais fácil é montar o diretório original em um idêntico no contêiner. Isso permite que você utilize o mesmo caminho durante a execução dentro do contêiner, por exemplo
```

    salmon index -t $PWD/data/ggal/transcriptome.fa -i transcript-index
```

Ou definir uma pasta que você queira montar como uma variável de ambiente, chamada `DATA`:

```bash
DATA=/workspace/gitpod/nf-training/data

> docker run --volume $DATA:$DATA --workdir $PWD my-image \

Reviewer: ```suggestion
docker run --volume $DATA:$DATA --workdir $PWD minha-imagem \
```

    salmon index -t $PWD/data/ggal/transcriptome.fa -i transcript-index
```

Agora cheque o conteúdo da pasta `transcript-index` utilizando o comando:

```bash
ls -la transcript-index
```

!!! note

    Note que as permissões para criação dos arquivos utilizado pelo Docker são `root`.

> ### Disponibilize o container no Docker Hub (bonus)

Reviewer: ```suggestion
### Disponibilize o contêiner no Docker Hub (bonus)
```

> Publique seu container no Docker Hub para compartilha-lo com outras pessoas.

Reviewer: ```suggestion
Publique seu contêiner no Docker Hub para compartilhá-lo com outras pessoas.
```

> Crie uma conta no site <https://hub.docker.com>. Entao no seu terminal shell execute o seguinte comando, utilizando seu usuário e senha que criou quando se registrou no Hub:

Reviewer: ```suggestion
Crie uma conta no site <https://hub.docker.com>. Então no seu terminal shell execute o seguinte comando, utilizando seu usuário e senha que criou quando se registrou no Hub:
```

```bash
docker login
```

Marque a imagem com seu nome de usuário Docker:

```bash

> docker tag my-image <user-name>/my-image

Reviewer: ```suggestion
docker tag minha-imagem <nome-de-usuario>/minha-imagem
```

```

Finalmente mande para o Docker Hub:

```bash

> docker push <user-name>/my-image

Reviewer: ```suggestion
docker push <nome-de-usuario>/minha-imagem
```

```

> Depois qualquer um conseguira baixar a imagem utilizando o comando:

Reviewer: ```suggestion
Depois disso, qualquer um conseguirá baixar a imagem utilizando o comando:
```

```bash
docker pull <user-name>/my-image
```

Note que depois de uma operação push e pull, o Docker printa o número de registro do container, por exemplo:

```console
Digest: sha256:aeacbd7ea1154f263cda972a96920fb228b2033544c2641476350b9317dab266
Status: Downloaded newer image for nextflow/rnaseq-nf:latest
```

> Isso é um identificador imutável e único que pode ser usado para referênciar a imagem de container de uma forma única. Por exemplo:

Reviewer: ```suggestion
Isso é um identificador imutável e único que pode ser usado para referenciar a imagem de contêiner de uma forma única. Por exemplo:
```

```bash
docker pull nextflow/rnaseq-nf@sha256:aeacbd7ea1154f263cda972a96920fb228b2033544c2641476350b9317dab266
```

> ### Execute um script do Nextflow utilizando um container Docker

Reviewer: ```suggestion
### Execute um script do Nextflow utilizando um contêiner Docker
```

A maneira mais simples de rodar um script Nextflow é usando a opção de linha de comando `-with-docker`:

```bash
nextflow run script2.nf -with-docker my-image
```

> Como visto na última seção, você também pode configurar o arquivo config (`nextflow.config`) para selecionar qual container utilizar invés de ter que especificar como um argumento de linha de comando toda vez.

Reviewer: ```suggestion
Como visto na última seção, você também pode configurar o arquivo config (`nextflow.config`) para selecionar qual contêiner utilizar invés de ter que especificar como um argumento de linha de comando toda vez.
```

## Singularity

> [Singularity](http://singularity.lbl.gov) é um motor de contêinerização desenvolvido para trabalhar com computação de alta performace em centro de dados, onde geralmente o Docker não é permitido por motivos de restrições de segurança.

Reviewer: ```suggestion
[Singularity](http://singularity.lbl.gov) é um motor de contêinerização desenvolvido para trabalhar com computação de alta performance em centro de dados, onde geralmente o Docker não é permitido por motivos de restrições de segurança.
```

!!! warning

    O Singularity não irá funcionar com Gitpod. Se você quer testar essa seção, por favor faça localmente, ou em um cluster.

### Crie imagens do Singularity

Imagens do Singularity são criadas utilizando um arquivo `Singularity` de uma forma similar ao Docker mas utilizando uma sintaxe diferente.

```singularity
Bootstrap: docker
From: debian:stretch-slim

%environment
export PATH=$PATH:/usr/games/

%labels
AUTHOR <seu nome>

%post

apt-get update && apt-get install -y locales-all curl cowsay
curl -sSL https://github.com/COMBINE-lab/salmon/releases/download/v1.0.0/salmon-1.0.0_linux_x86_64.tar.gz | tar xz \
&& mv /salmon-*/bin/* /usr/bin/ \
&& mv /salmon-*/lib/* /usr/lib/
```

Uma vez que você salvou o arquivo `Singularity`. Você pode criar uma imagem utilizando esses comandos:

```bash
sudo singularity build minha-imagem.sif Singularity
```

Nota: O comando `build` requer permissões `sudo`. Uma forma de contornar isso consiste em criar a imagem em uma estação de trabalho local e então implementar no cluster copiando o arquivo imagem.

> ### Executando um container

Reviewer: ```suggestion
### Executando um contêiner
```

```bash
touch ola.txt
ls -la
```

!!! info

    Note como os arquivos do sistema de arquivos hospedeiro são mostrados. O Singularity automaticamente monta o diretório do hospedeiro `$HOME` e usa como diretório de trabalho.

### Importe uma imagem do Docker

Uma forma mais fácil de criar um contêiner com o Singularity não necessitando da permissão `sudo` e melhorando a interoperabilidade dos contêineres é importando uma imagem de contêiner do Docker puxando diretamente do registro do Docker. Por exemplo:

```bash
singularity pull docker://debian:stretch-slim
```

O comando acima automaticamente baixa uma imagem Docker do Debian e converte para uma imagem Singularity no diretório atual com o nome`debian-jessie.simg`.

### Execute um script do Nextflow utilizando um contêiner Singularity

O Nextflow permite um uso de contêineres Singularity de forma tão fácil e transparente quanto com o Docker.

Simplesmente ative o uso do motor Singularity no lugar do Docker no arquivo de configurações do Nextflow utilizando a opção de linha de comando `-with-singularity`:

```bash
nextflow run script7.nf -with-singularity nextflow/rnaseq-nf
```

> Como antes, o contêiner Singularity também pode ser disponibilizado no arquivo de confirguração do Nextflow. Nós iremos ver como funciona isso mais tarde.

Reviewer: ```suggestion
Como antes, o contêiner Singularity também pode ser disponibilizado no arquivo de configuração do Nextflow. Nós iremos ver como funciona isso mais tarde.
```

Isso economiza tempo ao montar um ambiente conda toda vez que quiser utilizá-lo (como delineado nas seções anteriores).

Para fazer isso, você simplesmente precisa de um `Dockerfile` e utilizar o micromamba para instalar os pacotes. Porém, uma boa prática é ter o arquivo de receita YAML como nas seções anteriores, então nós iremos fazer isso aqui também, utilizando o mesmo `env.yml` utilizado anteriormente.

```yaml
--8<-- "nf-training/env.yml"
```

Então, nós podemos escrever nosso Dockerfile com o micromamba instalando os pacotes por esse arquivo de receita.

```dockerfile
FROM mambaorg/micromamba:0.25.1

LABEL image.author.name "Seu Nome Aqui"
LABEL image.author.email "seu@email.aqui"

COPY --chown=$MAMBA_USER:$MAMBA_USER env.yml /tmp/env.yml

RUN micromamba create -n nf-tutorial

RUN micromamba install -y -n nf-tutorial -f /tmp/env.yml && \
    micromamba clean --all --yes

ENV PATH /opt/conda/envs/nf-tutorial/bin:$PATH
```

O `Dockerfile` acima pega a imagem pai _mambaorg/micromamba_, e instala um ambiente `conda` utilizando `micromamba`, e então instala o `salmon`, o `fastqc` e o `multiqc`.

> Tente executar o pipeline RNA-seq de antes (script7.nf). Comece montando seu próprio `Dockerfile` micromamba (como mostrado acima), salve no seu repositório no Docker Hub, e oriente o Nextflow a rodar por esse container (mudando seu `nextflow.config`).

Reviewer: ```suggestion
Tente executar o pipeline RNA-seq de antes (script7.nf). Comece montando seu próprio `Dockerfile` micromamba (como mostrado acima), salve no seu repositório no Docker Hub, e oriente o Nextflow a rodar por esse contêiner (mudando seu `nextflow.config`).
```

!!! warning

>     Montar um contêiner Docker e disponibilizar no seu repósitorio pessoal pode levar &gt;10 minutos.

Reviewer: ```suggestion
    Montar um contêiner Docker e disponibilizar no seu repositório pessoal pode levar &gt;10 minutos.
```

??? example "Para um resumo dos passos a tomar, clique aqui:"

    1. crie um arquivo chamado `Dockerfile` no diretório atual (com os código acima).

    2. Monte a imagem: `docker build -t my-image .` (não esqueça o _._).

>     3. Publique a imagem docker na sua conta docker.

Reviewer: ```suggestion
    3. Publique a imagem Docker na sua conta do Docker Hub.
```

>         Algo parecido como o seguinte, com `<myrepo>` substituido para seu próprio Docker ID, sem _&lt;_ e _&gt;_ caracteres!

Reviewer: ```suggestion
        Algo parecido como o seguinte, com `<meurepo>` substituido para seu próprio ID do Docker, sem _&lt;_ e _&gt;_ caracteres!
```

>         Algo parecido como o seguinte, com `<meurepo>` substituido para seu próprio ID do Docker, sem _&lt;_ e _&gt;_ caracteres!

Reviewer: ```suggestion
        Algo parecido como o seguinte, com `<meurepo>` substituído para seu próprio ID do Docker, sem _&lt;_ e _&gt;_ caracteres!
```

>         `my-image` pode ser substituido por qualquer nome que escolher. como boa prática, escolha algo memorável e certifique que o nome combine com o nome usado no comando anterior.

Reviewer: ```suggestion
        `minha-imagem` pode ser substituído por qualquer nome que você escolher. Como boa prática, escolha algo memorável e certifique-se de que o nome combine com o nome usado no comando anterior.
```

        ```bash
        docker login

>         docker tag my-image <myrepo>/my-image

Reviewer: ```suggestion
        docker tag my-image <meurepo>/minha-imagem
```

>         docker push <myrepo>/my-image

Reviewer: ```suggestion
        docker push <meurepo>/minha-imagem
```

        ```

>     4. Adicione o arquivo imagem no arquivo `nextflow.config`.

Reviewer: ```suggestion
    4. Adicione o arquivo da imagem no arquivo `nextflow.config`.
```

>         ex. remova o seguinte de `nextflow.config`:

Reviewer: ```suggestion
        ex. remova o seguinte do arquivo `nextflow.config`:
```

        ```groovy
        process.container = 'nextflow/rnaseq-nf'
        ```

>         e mude por:

Reviewer: ```suggestion
        e adicione:
```

        ```groovy
        process.container = '<myrepo>/my-image'
        ```

>     5. Tente executar o Nextflow, ex.:

Reviewer: ```suggestion
    5. Tente executar o Nextflow, por exemplo:
```

        ```bash
        nextflow run script7.nf -with-docker
        ```

    Nextflow deve conseguir achar `salmon` pra rodar o processo.

## BioContainers

> Outro útilo recurso para conectar Bioconda e containers é o projeto [BioContainers](https://biocontainers.pro). BioContainers é uma inciativa da comunidade que prover um registro de imagens de container para toda receita Bioconda.

Reviewer: ```suggestion
Outro recurso útil para conectar Bioconda e contêineres é o projeto [BioContainers](https://biocontainers.pro). BioContainers é uma iniciativa da comunidade para prover um registro de imagens de contêiner para cada receita do Bioconda.
```

> Até agora, nós vimos como instalar pacotes com conda and micromamba, ambos localmente e com container. Com BioContainers, você não precisa criar sua própria imagem container pra as ferramentas que você quer, e não precisa utilizar conda or micromamba para instalar pacotes. Ele já disponiliza uma imagem Docker contendo os programas que você quer instalado. Por exemplo, você pode adquirir a imagem container do fastqc utilizando BioContainers:

Reviewer: ```suggestion
Até agora, nós vimos como instalar pacotes com conda e micromamba, ambos localmente e com contêiner. Com o BioContainers, você não precisa criar sua própria imagem de contêiner para as ferramentas que você quiser, e não precisa utilizar conda or micromamba para instalar pacotes. Ele já disponiliza uma imagem Docker contendo os programas que você quer instalado. Por exemplo, você pode adquirir a imagem de contêiner do fastqc utilizando BioContainers:
```

> Até agora, nós vimos como instalar pacotes com conda e micromamba, ambos localmente e com contêiner. Com o BioContainers, você não precisa criar sua própria imagem de contêiner para as ferramentas que você quiser, e não precisa utilizar conda or micromamba para instalar pacotes. Ele já disponiliza uma imagem Docker contendo os programas que você quer instalado. Por exemplo, você pode adquirir a imagem de contêiner do fastqc utilizando BioContainers:

Reviewer: ```suggestion
Até agora, nós vimos como instalar pacotes com conda e micromamba, ambos localmente e com contêiner. Com o BioContainers, você não precisa criar sua própria imagem de contêiner para as ferramentas que você quiser, e não precisa utilizar conda or micromamba para instalar pacotes. O BioContainers já disponibiliza uma imagem Docker contendo os programas que você quer instalado. Por exemplo, você pode adquirir a imagem de contêiner do fastqc utilizando BioContainers:
```

```bash
docker pull biocontainers/fastqc:v0.11.5
```

Você pode checar o registro dos pacotes que quer no [site oficial do BioContainers](https://biocontainers.pro/registry).

> Contrariamente a outros registros que irão puxar a ultima imagem quando nenhuma tag (version) é disponibilizada, você precisa especificar uma tag quando baixando do BioContainers (depois de dois pontos `:`, ex. fastqc:v0.11.5). Cheque as tags com o registro e escolha a que melhor se adequa a suas necessidades.

Reviewer: ```suggestion
Contrariamente a outros registros que irão puxar a última imagem quando nenhum rótulo (versão) é especificado, você precisa especificar um rótulo quando for baixar do BioContainers (depois de dois pontos `:`, por exemplo fastqc:v0.11.5). Cheque os rótulos com o registro e escolha a que melhor se adequa a suas necessidades.
```

> Contrariamente a outros registros que irão puxar a última imagem quando nenhum rótulo (versão) é especificado, você precisa especificar um rótulo quando for baixar do BioContainers (depois de dois pontos `:`, por exemplo fastqc:v0.11.5). Cheque os rótulos com o registro e escolha a que melhor se adequa a suas necessidades.

Reviewer: ```suggestion
Contrariamente a outros registros que irão puxar a última imagem quando nenhum rótulo (versão) é especificado, você precisa especificar um rótulo quando for baixar do BioContainers (depois de dois pontos `:`, por exemplo fastqc:v0.11.5). Cheque os rótulos com o registro e escolha o que melhor se adéqua a suas necessidades.
```

!!! tip

>     Você pode ter definições mais complexas dentro de seu bloco de processo deixando a imagem de container apropriada ou o pacote conda seria usada dependendo se o usuário seleciona singularity, Docker ou conda. Você pode clicar [aqui](https://nf-co.re/docs/contributing/modules#software-requirements) para mais informações e [aqui](https://github.com/nf-core/modules/blob/61f68913fefc20241ceccb671b104230b2d775d7/modules/bowtie2/align/main.nf#L6-L9) para um exemplo.

Reviewer: ```suggestion
    Você pode ter definições mais complexas dentro de seu bloco de processo deixando a imagem de contêiner apropriada ou o pacote conda para serem usadas dependendo se o usuário selecionou singularity, Docker ou conda. Você pode clicar [aqui](https://nf-co.re/docs/contributing/modules#software-requirements) para mais informações e [aqui](https://github.com/nf-core/modules/blob/61f68913fefc20241ceccb671b104230b2d775d7/modules/bowtie2/align/main.nf#L6-L9) para um exemplo.
```

### :material-progress-question: Exercises

!!! exercise

>     Durante o anterior tutorial RNA-Seq (script2.nf), nós criamos um índice. Dado que nós não temos salmom instalado localmente na maquina provida pelo Gitpod, nós temos que ou executar com `-with-conda` ou `-with-docker`. Sua tarefa agora é executar novamente com `-with-docker`, mas sem ter que criar sua própria imagem de container Docker. Invés disso, use a imagem BioContainers para salmon 1.7.0.

Reviewer: ```suggestion
    Durante a seção onde construímos o pipline de RNA-Seq, nós criamos um índice (script2.nf). Dado que nós não temos Salmon instalado localmente na máquina provida pelo Gitpod, nós temos que ou executar com `-with-conda` ou `-with-docker`. Sua tarefa agora é executar novamente com `-with-docker`, mas sem ter que criar sua própria imagem de contêiner Docker. Invés disso, use a imagem do BioContainers para Salmon 1.7.0.
```

>     Durante a seção onde construímos o pipline de RNA-Seq, nós criamos um índice (script2.nf). Dado que nós não temos Salmon instalado localmente na máquina provida pelo Gitpod, nós temos que ou executar com `-with-conda` ou `-with-docker`. Sua tarefa agora é executar novamente com `-with-docker`, mas sem ter que criar sua própria imagem de contêiner Docker. Invés disso, use a imagem do BioContainers para Salmon 1.7.0.

Reviewer: ```suggestion
    Durante a seção onde construímos o pipeline de RNA-Seq, nós criamos um índice (script2.nf). Dado que nós não temos Salmon instalado localmente na máquina provida pelo Gitpod, nós temos que ou executar com `-with-conda` ou `-with-docker`. Sua tarefa agora é executar novamente com `-with-docker`, mas sem ter que criar sua própria imagem de contêiner Docker. Invés disso, use a imagem do BioContainers para Salmon 1.7.0.
```

    ??? result

        ```bash
        nextflow run script2.nf -with-docker quay.io/biocontainers/salmon:1.7.0--h84f40af_0
        ```

!!! exercise "Bonus Exercise"

    Mude as diretivas do processo no `script5.nf` ou no arquivo `nextflow.config` para fazer o pipeline utilizar automaticamente BioContainers quando usando salmon, ou fastqc.

    !!! tip "Dica"

        temporariamente comente a linha `#!groovy process.container = 'nextflow/rnaseq-nf'` no arquivo `nextflow.config` para ter certeza que o processo está utilizando BioContainers que você configurou, e não a imagem de container que estavamos usando durante esse treinamento.

    ??? result

        Com essas mudanças, você deve ser capaz de executar o pipeline com BioContainers executando a seguinte linha de comando:

        ```bash
        nextflow run script5.nf
        ```

>         com as sequintes diretivas de container para cada processo:

Reviewer: ```suggestion
        com as sequintes diretivas de contêiner para cada processo:
```
